### DISCUSSION TITLE
CRSF and PWM on same RX

### DISCUSSION THREAD

**Discussion by MacDaddyFPV** [ExpressLRS/ExpressLRS#1330] [Ideas]
I have a PDB which I have designed with onboard MCU to do canvas mode OSD for HDZERO in a wing with a PWM RX. Problem is I can't send stick commands as there is no through connection from RX to VTX.

I would like if there could be the ability on a 5ch RX to set 2 channels to CRSF (TX/RX) and the remaining 3 to stay PWM. This would allow me to connect to the MCU (which is flashed with bf) to enable stick commands to be directed through MSP.

*Instead of commenting "I want this too", consider leaving a :+1: instead, to reduce spam and clutter*

**Comment by CapnBry** [comment]
I'm looking into this, it certainly is doable but requires some internal changes. I'm just trying to work out how it will be accessible because I don't want to create a whole new target for it.

**Reply by sercona** [reply]
this would be really useful.  my use-case might be to have a partial smart solution (pwm rx sending steering directly to the servo - not needing any cpu processing help - but then using crsf over serial (on that same rx) to a FC and have that do the throttle stuff for me.

more for debug I guess or to test latency thru my FC vs direct; but it would be nice to be able to use my surface pwm rx's in pure serial mode and/or allocate 2 wires for serial and the rest for pwm.

**Comment by danxdz** [comment]
How about one option from webui? Like in debug mode, serial keeps working..

**Comment by kolins-cz** [comment]
Hi, I managed to do this in 3.1.2 and it tested OK. I had to mess around various JSON and .ini files in ELRS codebase and one JSON file in configurator.

Basically I have added a new target

Most important change was this:
Generic 2400 PWMP666.json (made from Generic 2400 PWMP6.json) with Serial port defines
`{
	"serial_rx": 3,
    "serial_tx": 1,
    "radio_dio1": 4,
    "radio_miso": 12,
    "radio_mosi": 13,
    "radio_nss": 15,
    "radio_rst": 2,
    "radio_sck": 14,
    "power_min": 0,
    "power_high": 0,
    "power_max": 0,
    "power_default": 0,
    "power_control": 0,
    "power_values": [13],
    "led": 16,
    "pwm_outputs": [0,9,10,5],
    "vbat": 17,
    "vbat_offset": 12,
    "vbat_scale": 310
}`

I would like to have this in main codebase, but I do not know how to properly push pull request. Maybe dev can do this in like a minute or so?

**Reply by CapnBry** [reply]
> I would like to have this in main codebase, but I do not know how to properly push pull request. Maybe dev can do this in like a minute or so?

The  reason it doesn't exist is because it's a terrible way to do it (making a target for it). Having tons of slightly modified targets is just a mess. All you have to do is edit the hardware.html and from there you can make anything anything, so having to maintain everyone's "I just wanted to change a couple pins THIS way" just creates work for developers to maintain.

I'd much rather have it so you can just change the function of the pin from the main webui, selecting "CRSF TX" instead of e.g. "50Hz" as the mode. This would be done for all targets so even standard receivers could output PWM on one of their pins too without someone wanting to add "Just the EP2 target but the CRSF RX pin is PWM out" or any of the other million combinations.

Anyone who wants that configuration, just:
* Copy paste the JSON into a file, named hardware.json or something
* Go to the webui `/hardware.html` and drag and drop/upload the file at the top of the page. WARNING if your target is not the exact layout kolins-cz was using, this will not work properly. This is why it is better to just alter the pins for your hardware, since there is a lot of hardware out there and a lot of it doesn't have the pins connected the same way, which is why "just add another target!" doesn't work well.
* Save and reboot.

**Reply by kolins-cz** [reply]
>     * Copy paste the JSON into a file, named hardware.json or something
>     * Go to the webui `/hardware.html` and drag and drop/upload the file at the top of the page. 
>     * Save and reboot.

This is what I hoped for to find in documentation. Or better as an option in web configurator.
Do I understand that general pinout is defined run-time from json file? I thought that it is compiled in.

**Reply by pkendall64** [reply]
> > ```
> > * Copy paste the JSON into a file, named hardware.json or something
> > * Go to the webui `/hardware.html` and drag and drop/upload the file at the top of the page. 
> > * Save and reboot.
> > ```
> 
> This is what I hoped for to find in documentation. Or better as an option in web configurator. Do I understand that general pinout is defined run-time from json file? I thought that it is compiled in.

Yes, it's all defined in the JSON file for ESP based targets since version 3.0.

**Comment by pkendall64** [comment]
I actually have a branch in progress that does this in a much more configurable way. You can enable serial protocol on appropriate PWM pins, and even select the protocol on the pins i.e. CRSF/SBUS as well as inverted. https://github.com/pkendall64/ExpressLRS/tree/rx-serial-refactor

**Reply by rotorgeeks** [reply]
Please forgive my ignorance but how would I go about loading and using this?  It's exactly what I've been looking for.

**Reply by pkendall64** [reply]
At the moment you would have to compile it with platformio and upload from there too.

**Reply by rotorgeeks** [reply]
I think I'm close.  I got it uploaded to the RX and see the modified mode description that includes CRSF and SBUS but it's not showing as an option in the mode pulldown boxes.  Does it require any special user defines?
Running it on a Matek R24-P (5ch PWM receiver)
-edit- Just realized the Matek description says this RX is locked at 50Hz update rate.  I'll try tomorrow with the newer R24-P6 and see if that opens it up.

**Reply by rotorgeeks** [reply]
I'm getting the same result on the Matek R24-P6 (target DIY 2.4 GHz / DIY 2400 RX PWMP EX)
The Web GUI doesn't show the option in output mode.  LUA script shows Protocol CRSF.  In Output mapping I can choose an output mode of Serial TX or Serial RX but it doesn't stick after selection, it reverts back to the previous value.  Any advice?

**Reply by pkendall64** [reply]
Only some of the pins allow SerialTX/RX to be configured, if they jump back when you select them then it's can't do Serial on that pin.
So, look at the RX itself and only choose the pins marked as RX or TX as serial pins.

**Reply by rotorgeeks** [reply]
Brilliant, thank you!  It's working great now.

**Reply by rotorgeeks** [reply]
Did this make it into a release yet?

**Reply by pkendall64** [reply]
It's in the master branch and will be in the 3.3 release "real soo now"
